File: DSA/src/Questions/Array/MergeInterval.java
// File: questions/MergeInterval.java
package Questions.Array;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class MergeInterval {
    static int[][] merge(int[][] intervals) {
        Arrays.sort(intervals, (a, b) -> {
            if (a[0] == b[0]) {
                return b[1] - a[1];
            }
            return a[0] - b[0];
        });
        List<int[]> result = new ArrayList<>();

        for (int[] interval : intervals) {
            if (result.isEmpty()) {
                result.add(interval);
            } else {
                // Pointing to the last interval from the result
                int[] lastInterval = result.get(result.size() - 1);

                if (lastInterval[1] >= interval[0]) {
                    lastInterval[1] = Math.max(lastInterval[1], interval[1]);
                } else {
                    result.add(interval);
                }
            }
        }
        return result.toArray(new int[result.size()][]);
    }

    public static void main(String[] args) {
        int[][] intervals = new int[][] { { 1, 3 }, { 2, 6 }, { 8, 10 }, { 15, 18 } };
        int[][] result = merge(intervals);
        for (int[] interval : result) {
            System.out.println(Arrays.toString(interval));
        }
    }
}

File: DSA/Python/Questions/Array/MergeInterval.py
"""
# Question: Merge Intervals
# Link: https://leetcode.com/problems/merge-intervals/

# Given an array of intervals where intervals[i] = [starti, endi],
# merge all overlapping intervals and return the non-overlapping intervals.

# Time Complexity: O(n log n) due to sorting
# Space Complexity: O(n) for storing result

# Algorithm:
# 1. Sort intervals based on start time
# 2. Iterate through sorted intervals
# 3. Merge overlapping intervals by updating end time
# 4. Add non-overlapping intervals to result

# Key Components:
# - merge(): Main function implementing interval merging
# - Sorting with custom lambda for interval comparison
# - Dynamic result building with overlap checking
"""

from typing import List


class MergeInterval:
    def merge(self, intervals: List[List[int]]) -> List[List[int]]:
        # Sort intervals based on start time
        intervals.sort(key=lambda l: l[0])
        merged = []
        for interval in intervals:
            if not merged or interval[0] > merged[-1][1]:
                merged.append(interval)
            else:
                merged[-1][1] = max(merged[-1][1], interval[1])

        return merged


def main():
    solution = MergeInterval()
    intervals = [[1, 3], [2, 6], [8, 10], [15, 18]]
    result = solution.merge(intervals)
    print(result)  # Expected output: [[1, 6], [8, 10], [15, 18]]


if __name__ == "__main__":
    main()
